clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -227,9 +234,83 @@ public void changePartitionAssignment(List<TopicPartition> assignments) {
             if (!this.subscribedTopics.contains(tp.topic()))
                 throw new IllegalArgumentException("Assigned partition " + tp + " for non-subscribed topic.");
         this.clearAssignment();
-        this.assignedPartitions.addAll(assignments);
+        for (TopicPartition tp: assignments)
+            this.assignedPartitions.put(tp, new TopicPartitionState());
         this.needsPartitionAssignment = false;
     }
 
+    public boolean isAssigned(TopicPartition tp) {
+        return assignedPartitions.containsKey(tp);
+    }
+
+    public boolean isPaused(TopicPartition tp) {
+        return assertAssignment(tp).paused;
+    }
+
+    public boolean isFetchable(TopicPartition tp) {
+        return fetchable.contains(tp);
+    }
+
+    public void pause(TopicPartition tp) {
+        assertAssignment(tp).pause();
+    }
+
+    public void resume(TopicPartition tp) {
+        assertAssignment(tp).resume();
+    }
+
+    private static class TopicPartitionState {
+        private Long consumed;   // offset exposed to the user
+        private Long fetched;    // current fetch position
+        private Long committed;  // last committed position
+
+        private boolean paused;  // whether this partition has been paused by the user
```

Comment:
@guozhangwang That's actually close to my initial intention and it would definitely simplify the code. I think the only downside is that the call to hasAllFetchPositions (which is called on every poll()) would require a scan over all the assigned partitions. Do you think this collection would ever get large enough for that to matter?